### DISCUSSION TITLE
Unable to upload files

### DISCUSSION THREAD

**Discussion by christian-lorenz** [thecodingmachine/graphqlite#391] [Q&A]
Symfony 5, apollo (vue-apollo) / vue 2, PHP 7.4

I installed `ecodev/graphql-upload` but needed to set it to v5, as 6 requires a newer webonyx/graphql-php or so. (And somehow I can't install graphql 4.3 - composer says 4.1 is the latest)

I always get:  
cannot find GraphQL type "File". Check your TypeMapper configuration.
vendor/thecodingmachine/graphqlite/src/Mappers/CannotMapTypeException.php (line 49)

Apollo config:
```
const apolloClient = new ApolloClient({
  link: createUploadLink({ uri: 'http://event-backend.devs/graphql' }),
  cache: new InMemoryCache()
})

Vue.use(VueApollo);

const apolloProvider = new VueApollo({
  defaultClient: apolloClient,
})
```

Mutation is like
```#import "./FileFragment.gql"

mutation uploadFile ($file: Upload!) {
  singleUpload (file: $file) {
    ...file
  }
}

 <ApolloMutation
        :mutation="require('../../graphql/uploadFile.gql')"
        :variables="{
         file: eventSeries.logo
      }">
      
      <v-file-input
                  v-model="eventSeries.logo"
                  :rules="rulesImages"
                  accept="image/png, image/jpeg"
                  prepend-icon="mdi-camera"
                  label="Logo"
              ></v-file-input>
```

Queries and so on work fine. What am I missing?

**Comment by oojacoboo** [comment]
Are you using the Symfony bundle?  It's still targeting graphqlite 4.1.x.

**Reply by oojacoboo** [reply]
Also, regarding the error, did you add the `UploadMiddleware` to your middleware pipe?

**Reply by christian-lorenz** [reply]
Uhm... I think I created the symfony app either by the symfony new command itself or by composer.
Would a composer.json help?

And I didn't do anything regarding the middleware.

**Reply by christian-lorenz** [reply]
```
{
  "type": "project",
  "license": "proprietary",
  "minimum-stability": "stable",
  "prefer-stable": true,
  "require": {
    "php": ">=7.4",
    "ext-ctype": "*",
    "ext-curl": "*",
    "ext-iconv": "*",
    "ext-json": "*",
    "doctrine/annotations": "^1.0",
    "doctrine/doctrine-bundle": "^2.1",
    "doctrine/doctrine-migrations-bundle": "^3.0",
    "doctrine/orm": "^2.7",
    "ecodev/graphql-upload": "5.*",
    "symfony/apache-pack": "^1.0",
    "symfony/console": "5.3.*",
    "symfony/dotenv": "5.3.*",
    "symfony/flex": "^1.3.1",
    "symfony/framework-bundle": "5.3.*",
    "symfony/http-client": "5.3.*",
    "symfony/runtime": "5.3.*",
    "symfony/webpack-encore-bundle": "1.11.*",
    "symfony/yaml": "5.3.*",
    "thecodingmachine/graphqlite-bundle": "^4.1"
  },
  "require-dev": {
    "symfony/maker-bundle": "^1.33"
  },
  "config": {
    "optimize-autoloader": true,
    "preferred-install": {
      "*": "dist"
    },
    "sort-packages": true
  },
  "autoload": {
    "psr-4": {
      "App\\": "src/"
    }
  },
  "autoload-dev": {
    "psr-4": {
      "App\\Tests\\": "tests/"
    }
  },
  "replace": {
    "symfony/polyfill-ctype": "*",
    "symfony/polyfill-iconv": "*",
    "symfony/polyfill-php72": "*"
  },
  "scripts": {
    "auto-scripts": {
      "cache:clear": "symfony-cmd",
      "assets:install %PUBLIC_DIR%": "symfony-cmd"
    },
    "post-install-cmd": [
      "@auto-scripts"
    ],
    "post-update-cmd": [
      "@auto-scripts"
    ]
  },
  "conflict": {
    "symfony/symfony": "*"
  },
  "extra": {
    "symfony": {
      "allow-contrib": true,
      "require": "5.3.*"
    }
  }
}
```
composer.json

**Reply by oojacoboo** [reply]
`"thecodingmachine/graphqlite-bundle": "^4.1"`

That's the Symfony bundle.  This is the base library that bundle uses.  I cannot help you with that.  That repo needs target 4.3.  So, you'll want to take the issue there.

As for the middleware, you'll need to add that to process the request before handing it off to GraphQLite for processing.

**Reply by christian-lorenz** [reply]
Well, I think I can't just tell "composer: there's a 4.3 version - just try harder".
I mean, I made a basic (ok, haha, as webpack encore doesn't work since a month and needs some treatment as well) installation of symfony and followed your documentation to make file uploads possible. :-/

Can you provide me a working composer.json or a symfony skeleton? Because I really don't know where / what to do here.

**Reply by oojacoboo** [reply]
If you do not understand what's being said below, research every bit of it, then return with questions.

The `graphqlite-bundle` is a Symfony bundle.  It is its own repository.  It's targeting version 4.1 of this repository.  In short, it needs to be updated.  You can fork it and do so, or you can submit a PR there, or you can submit an issue to have the maintainers do so.

**Reply by christian-lorenz** [reply]
Hi man, please don't feel offended.
I installed symfony the common way.
I added graphqlite like you wrote on the docs.
I installed the upload things (and even went beyond and configured apollo I hope the right way).

And you're telling me: I did it wrong.
Usually you tag releases (pushes) and dev-master points to the latest one. Obviously you didn't here (for whatever reason - maybe because of release candidates or whatever)

But please don't tell me, I did something wrong / don't understand enough.
I'm a common dev working in a company which I think is your target group with graphqlite.

So please tell me:
Is file-upload working out of the box on how you descripted it in your docs?
If yes: Please send me your composer.json
If not: Please tell me what you needed to be done that it worked.

**Reply by oojacoboo** [reply]
The documentation for these framework adapters is still part of the main GraphQLite docs.  Please see the following issue ticket:

https://github.com/thecodingmachine/graphqlite/issues/376

This repository and any adapters that adapt this repository to work with another framework: Symfony, Laravel, etc. are being maintained separately.

File uploading works with GraphQLite.  We use it at our company.  I've already told you what needs to be done.  I don't know enough about your code design to tell you exactly where to put it.  But you need to "pre-process" the request before handing it off to GraphQLite.

**Reply by christian-lorenz** [reply]
I'm still getting "cannot find GraphQL type "File". Check your TypeMapper configuration."
And I'm running out of ideas :(

I tried it like this:

```php
<?php


namespace App\EventSubscriber;

use GraphQL\Upload\UploadMiddleware;
use Nyholm\Psr7\Factory\Psr17Factory;
use Symfony\Bridge\PsrHttpMessage\Factory\HttpFoundationFactory;
use Symfony\Bridge\PsrHttpMessage\Factory\PsrHttpFactory;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpKernel\Event\ControllerEvent;
use Symfony\Component\HttpKernel\KernelEvents;
use TheCodingMachine\Graphqlite\Bundle\Controller\GraphqliteController;

class GraphQLUploadSubscriber implements EventSubscriberInterface
{
    private GraphqliteController $graphqliteController;

    public function __construct(GraphqliteController $graphqliteController)
    {
        $this->graphqliteController = $graphqliteController;
    }

    /**
     * @param ControllerEvent $event
     *
     * @return Response|void|null
     */
    public function onKernelController(ControllerEvent $event)
    {
        $controller = $event->getController();

        if (!is_array($controller) || !$controller[0] instanceof GraphqliteController) {
            return;
        }

        $symfonyRequest = $event->getRequest();

        $psr17Factory = new Psr17Factory();
        $psrHttpFactory = new PsrHttpFactory($psr17Factory, $psr17Factory, $psr17Factory, $psr17Factory);
        $uploadMiddleware = new UploadMiddleware();
        $httpFoundationFactory = new HttpFoundationFactory();

        // convert to PSR request
        $psrRequest = $psrHttpFactory->createRequest($symfonyRequest);
        $psrRequest = $uploadMiddleware->processRequest($psrRequest);
        
        // convert it back
        $symfonyRequest = $httpFoundationFactory->createRequest($psrRequest);

        try {
            return $this->graphqliteController->handleRequest($symfonyRequest);
        } catch (\Exception $e) {
            // This one still throws with "cannot find GraphQL type "File". Check your TypeMapper configuration."
            die($e->getMessage());
        }
    }

    /**
     * @return array
     */
    public static function getSubscribedEvents(): array
    {
        return [
            KernelEvents::CONTROLLER => 'onKernelController',
        ];
    }
}
```

**Reply by christian-lorenz** [reply]
Ok, forget it - the error was in the apollo request itself. You don't even need the Subscriber D:
